### PR TITLE
Add retry to get pod logs

### PR DESCRIPTION
**Pull request by mauriciopoppe**
#### What type of PR is this?

/kind bug
/sig-storage
/cc @msau42 
/assign @oomichi

#### What this PR does / why we need it:
Adds a retry in the e2e framework's `MatchContainerOutput` function to get pod logs. I'll backport it to 1.21

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/framework/util.go [hunk 1]
```diff
@@ -876,9 +876,21 @@ func (f *Framework) MatchContainerOutput(
...
+		func() (bool, error) {
+			logs, podLogErr = e2epod.GetPodLogs(f.ClientSet, ns, podStatus.Name, containerName)
+			if err == nil {
+				return true, nil
+			}
+			return false, nil
+		})
 	if err != nil {
+		err = fmt.Errorf("%w: %w", err, podLogErr)
 		Logf("Failed to get logs from node %q pod %q container %q. %v",
```
Comment (yujuhong) [inline test/e2e/framework/util.go:894]: Can we also log the pod status here? The pod gets deleted right after the failure, and we have no way to see this information.

File: test/e2e/framework/util.go [hunk 2]
```diff
@@ -876,9 +876,21 @@ func (f *Framework) MatchContainerOutput(
...
+		1*time.Minute,
+		func() (bool, error) {
+			logs, podLogErr = e2epod.GetPodLogs(f.ClientSet, ns, podStatus.Name, containerName)
+			if err == nil {
+				return true, nil
+			}
+			return false, nil
+		})
 	if err != nil {
+		err = fmt.Errorf("%w: %w", err, podLogErr)
```
Comment (oomichi) [inline test/e2e/framework/util.go:893]: Need to fix here due to the error:
```
test/e2e/framework/util.go:893:9: fmt.Errorf call has more than one error-wrapping directive %w
```
as https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/109578/pull-kubernetes-verify/1516918382241255424

File: test/e2e/framework/util.go [hunk 3]
```diff
@@ -876,9 +876,21 @@ func (f *Framework) MatchContainerOutput(
...
-	logs, err := e2epod.GetPodLogs(f.ClientSet, ns, podStatus.Name, containerName)
+	// Sometimes the actual containers take a few seconds to get started, try to get logs for 1m.
+	var logs string
+	var podLogErr error
+	err = wait.PollImmediate(
+		2*time.Second, /* from pods/wait.go:poll */
+		1*time.Minute,
+		func() (bool, error) {
+			logs, podLogErr = e2epod.GetPodLogs(f.ClientSet, ns, podStatus.Name, containerName)
+			if err == nil {
```
Comment (oomichi) [inline test/e2e/framework/util.go:887]: Doesn't here need to check `podLogErr` as a returned value of `e2epod.GetPodLogs()` instead of `err`?

**Review by oomichi — COMMENTED** [review]
/cc @oomichi

File: test/e2e/framework/util.go (hunk 3 above)
Comment (mauriciopoppe) [reply test/e2e/framework/util.go:887]: thanks for catching this! I think I started with `err` but changed it later to `podLogErr` to join both errors.

**Comment by mauriciopoppe** [comment]
/hold
I'm making this change internally first, I'll update this PR later once we know that the fix works.

File: test/e2e/framework/util.go (hunk 2 above)
Comment (mauriciopoppe) [reply test/e2e/framework/util.go]: thanks, fixed.

**Comment by aojea** [comment]
@mauriciopoppe this was working and solid for us if you want to borrow some ideas

https://github.com/kubernetes/kubernetes/blob/f173d01c011c3574dea73a6fa3e20b0ab94531bb/test/e2e/network/conntrack.go#L84-L96

**Pull request closed without merging**